backend/src/schema/resolvers/passwordReset.spec.js

```diff
@@ -16,158 +21,202 @@ const getAllPasswordResets = async () => {
+        it('resolves anyways', async () => {
+          await expect(mutate({ mutation, variables })).resolves.toMatchObject({
+            data: { requestPasswordReset: true },
+          })
+        })
+
+        it('creates no node', async () => {
+          await mutate({ mutation, variables })
+          const resets = await getAllPasswordResets()
+          expect(resets).toHaveLength(0)
+        })
       })
-    })
 
-    describe('with a valid email', () => {
-      const variables = { email: 'user@example.org' }
+      describe('with a valid email', () => {
+        beforeEach(() => {
+          variables = { ...variables, email: 'user@example.org' }
+        })
 
-      it('resolves', async () => {
-        await expect(client.request(mutation, variables)).resolves.toEqual({
-          requestPasswordReset: true,
+        it('resolves', async () => {
+          await expect(mutate({ mutation, variables })).resolves.toMatchObject({
+            data: { requestPasswordReset: true },
+          })
         })
-      })
 
-      it('creates node with label `PasswordReset`', async () => {
-        await client.request(mutation, variables)
-        const resets = await getAllPasswordResets()
-        expect(resets).toHaveLength(1)
-      })
+        it('creates node with label `PasswordReset`', async () => {
+          let resets = await getAllPasswordResets()
+          expect(resets).toHaveLength(0)
+          await mutate({ mutation, variables })
+          resets = await getAllPasswordResets()
+          expect(resets).toHaveLength(1)
+        })
 
-      it('creates a reset code', async () => {
-        await client.request(mutation, variables)
-        const resets = await getAllPasswordResets()
-        const [reset] = resets
-        const { code } = reset.properties
-        expect(code).toHaveLength(6)
+        it('creates a reset nonce', async () => {
+          await mutate({ mutation, variables })
+          const resets = await getAllPasswordResets()
+          const [reset] = resets
+          const { nonce } = reset.properties
+          expect(nonce).toHaveLength(6)
+        })
       })
     })
   })
+})
+
+describe('resetPassword', () => {
+  const setup = async (options = {}) => {
+    const { email = 'user@example.org', issuedAt = new Date(), nonce = 'abcdef' } = options
 
-  describe('resetPassword', () => {
-    const setup = async (options = {}) => {
-      const { email = 'user@example.org', issuedAt = new Date(), code = 'abcdef' } = options
+    const session = driver.session()
+    await createPasswordReset({ driver, email, issuedAt, nonce })
+    session.close()
+  }
 
-      const session = driver.session()
-      await createPasswordReset({ driver, email, issuedAt, code })
-      session.close()
+  const mutation = gql`
+    mutation($nonce: String!, $email: String!, $newPassword: String!) {
+      resetPassword(nonce: $nonce, email: $email, newPassword: $newPassword)
     }
+  `
+  beforeEach(() => {
```

Comment:
> [<img alt="mattwr18" height="40" width="40" align="left" src="https://avatars1.githubusercontent.com/u/38261864?s=88&v=4">](https://github.com/mattwr18) **Authored by [mattwr18](https://github.com/mattwr18)**
_<time datetime="2019-09-10T08:11:47Z" title="Tuesday, September 10th 2019, 10:11:47 am +02:00">Sep 10, 2019</time>_

---
> **Outdated (history rewrite)** - original diff
---
```diff
@@ -16,158 +21,203 @@ const getAllPasswordResets = async () => {
   return resets
 }
 
-describe('passwordReset', () => {
-  beforeEach(async () => {
-    client = new GraphQLClient(host)
-    await factory.create('User', {
-      email: 'user@example.org',
-      role: 'user',
-      password: '1234',
-    })
-  })
+beforeEach(() => {
+  variables = {}
+})
 
-  afterEach(async () => {
-    await factory.cleanDatabase()
+beforeAll(() => {
+  const { server } = createServer({
+    context: () => {
+      return {
+        driver,
+        neode,
+        user: authenticatedUser,
+      }
+    },
   })
+  mutate = createTestClient(server).mutate
+})
 
-  describe('requestPasswordReset', () => {
-    const mutation = `mutation($email: String!) { requestPasswordReset(email: $email) }`
+afterEach(async () => {
+  await factory.cleanDatabase()
+})
 
-    describe('with invalid email', () => {
-      const variables = { email: 'non-existent@example.org' }
+describe('passwordReset', () => {
+  describe('given a user', () => {
+    beforeEach(async () => {
+      await factory.create('User', {
+        email: 'user@example.org',
+      })
+    })
+
+    describe('requestPasswordReset', () => {
+      const mutation = gql`
+        mutation($email: String!) {
+          requestPasswordReset(email: $email)
+        }
+      `
 
-      it('resolves anyways', async () => {
-        await expect(client.request(mutation, variables)).resolves.toEqual({
-          requestPasswordReset: true,
+      describe('with invalid email', () => {
+        beforeEach(() => {
+          variables = { ...variables, email: 'non-existent@example.org' }
         })
-      })
 
-      it('creates no node', async () => {
-        await client.request(mutation, variables)
-        const resets = await getAllPasswordResets()
-        expect(resets).toHaveLength(0)
+        it('resolves anyways', async () => {
+          await expect(mutate({ mutation, variables })).resolves.toMatchObject({
+            data: { requestPasswordReset: true },
+          })
+        })
+
+        it('creates no node', async () => {
+          await mutate({ mutation, variables })
+          const resets = await getAllPasswordResets()
+          expect(resets).toHaveLength(0)
+        })
       })
-    })
 
-    describe('with a valid email', () => {
-      const variables = { email: 'user@example.org' }
+      describe('with a valid email', () => {
+        beforeEach(() => {
+          variables = { ...variables, email: 'user@example.org' }
+        })
 
-      it('resolves', async () => {
-        await expect(client.request(mutation, variables)).resolves.toEqual({
-          requestPasswordReset: true,
+        it('resolves', async () => {
+          await expect(mutate({ mutation, variables })).resolves.toMatchObject({
+            data: { requestPasswordReset: true },
+          })
         })
-      })
 
-      it('creates node with label `PasswordReset`', async () => {
-        await client.request(mutation, variables)
-        const resets = await getAllPasswordResets()
-        expect(resets).toHaveLength(1)
-      })
+        it('creates node with label `PasswordReset`', async () => {
+          let resets = await getAllPasswordResets()
+          expect(resets).toHaveLength(0)
+          await mutate({ mutation, variables })
+          resets = await getAllPasswordResets()
+          expect(resets).toHaveLength(1)
+        })
 
-      it('creates a reset code', async () => {
-        await client.request(mutation, variables)
-        const resets = await getAllPasswordResets()
-        const [reset] = resets
-        const { code } = reset.properties
-        expect(code).toHaveLength(6)
+        it('creates a reset nonce', async () => {
+          await mutate({ mutation, variables })
+          const resets = await getAllPasswordResets()
+          const [reset] = resets
+          const { nonce } = reset.properties
+          expect(nonce).toHaveLength(6)
+        })
       })
     })
   })
+})
+
+describe('resetPassword', () => {
+  const setup = async (options = {}) => {
+    const { email = 'user@example.org', issuedAt = new Date(), nonce = 'abcdef' } = options
 
-  describe('resetPassword', () => {
-    const setup = async (options = {}) => {
-      const { email = 'user@example.org', issuedAt = new Date(), code = 'abcdef' } = options
+    const session = driver.session()
+    await createPasswordReset({ driver, email, issuedAt, nonce })
+    session.close()
+  }
 
-      const session = driver.session()
-      await createPasswordReset({ driver, email, issuedAt, code })
-      session.close()
+  const mutation = gql`
+    mutation($nonce: String!, $email: String!, $newPassword: String!) {
+      resetPassword(nonce: $nonce, email: $email, newPassword: $newPassword)
     }
+  `
+  const nonce = 'abcdef'
```


why do you have a default and then set the same value to a variable?

Reply:
> [<img alt="mattwr18" height="40" width="40" align="left" src="https://avatars1.githubusercontent.com/u/38261864?s=88&v=4">](https://github.com/mattwr18) **Authored by [mattwr18](https://github.com/mattwr18)**
_<time datetime="2019-09-10T08:17:45Z" title="Tuesday, September 10th 2019, 10:17:45 am +02:00">Sep 10, 2019</time>_

---


so this calls `createPasswordReset` with the nonce `abcdef` as a default, then uses that same nonce to find the `PasswordReset` node, but if the user passes in a different nonce like `slkdjf`, then it will create a `PasswordReset` node with that `nonce`, but will not match the `resetPassword` nonce?

Reply:
> [<img alt="roschaefer" height="40" width="40" align="left" src="https://avatars1.githubusercontent.com/u/38261864?s=88&v=4">](https://github.com/roschaefer) **Authored by [roschaefer](https://github.com/roschaefer)**
_<time datetime="2019-09-10T14:22:23Z" title="Tuesday, September 10th 2019, 4:22:23 pm +02:00">Sep 10, 2019</time>_

---


That's confusing! I removed the `const nonce` and set it with a hard-coded string where I need it.

I don't like the `setup()`  pattern anymore, too, but keep it for now.